tcp.go

```diff
@@ -32,33 +32,55 @@ var defaultPortByService = map[string]int{
 
 var errorFailedToVerifyService = errors.New("Failed to verify service")
 
-func (fm *Frontman) runTCPCheck(addr *net.TCPAddr, hostname string, service string) (m MeasurementsMap, err error) {
-
+func (fm *Frontman) runTCPCheck(addr *net.TCPAddr, hostname string, service string) (MeasurementsMap, error) {
 	service = strings.ToLower(service)
 
+	// Check if we have to autodetect port by service name
 	if addr.Port <= 0 {
-		if port, exists := defaultPortByService[service]; exists {
-			addr.Port = port
-		} else {
-			err = fmt.Errorf("failed to auto-determine port for '%s'", service)
-			return
+		// Lookup service by default port
+		port, exists := defaultPortByService[service]
+		if !exists {
+			return nil, fmt.Errorf("failed to auto-determine port for '%s'", service)
 		}
+		addr.Port = port
 	}
 
 	prefix := fmt.Sprintf("net.tcp.%s.%d.", service, addr.Port)
 
-	m = MeasurementsMap{
+	// Initialise MeasurementsMap
+	m := MeasurementsMap{
 		prefix + "success": 0,
 	}
 
+	// Start measuring execution time
 	started := time.Now()
+	// Calculate execution time in the end
+	defer func() {
+		m[prefix+"connectTime_s"] = time.Since(started).Seconds()
+	}()
+
+	// Open connection to the specified addr
 	conn, err := net.DialTimeout("tcp", addr.String(), secToDuration(fm.NetTCPTimeout))
 	if err != nil {
-		return
+		return m, err
 	}
-
 	defer conn.Close()
 
+	// Execute the check
+	err = fm.executeCheck(conn, service, hostname)
+	if err != nil {
+		return m, fmt.Errorf("failed to verify '%s' service on %d port: %s", service, addr.Port, err.Error())
+	}
+
+	// Mark check as successfull
+	m[prefix+"success"] = 1
+
+	return m, nil
+}
+
+// executeCheck executes the check based on the passed service name
+func (fm *Frontman) executeCheck(conn net.Conn, service, hostname string) error {
```

Comment:
One more thing.  Frontman is the top-level entity and shared across different types of checks: ICMP, SSL, TCP, Web.

So now you split it into the 2 top-level methods.
1. I think you can remove the struct pointer here and just pass `fm.NetTCPTimeout` as arg
2. You should rename the method it to mention the type of check. Maybe  executeTCPCheck

Reply:
1. ✅ 
2. The function does not plainly check TCP. Inside the method there are differentiations which protocol is checked. The connection is also given to the function. Therefor I think adding TCP to the name could be a bit misleading 🤔

Reply:
TCP is a transport protocol the other ones are underlying protocols. So you can use smth like "checkUnderlyingTCProtocol"